xwiki-rendering-macros/xwiki-rendering-macro-html/src/main/java/org/xwiki/rendering/internal/macro/html/HTMLMacroXHTMLChainingRenderer.java

```diff
@@ -154,4 +157,37 @@ protected BlockStateChainingListener getBlockState()
     {
         return (BlockStateChainingListener) getListenerChain().getListener(HTMLMacroBlockStateChainingListener.class);
     }
+
+    /**
+     * @return a span element if we are inside an inline macro. Else a div.
+     */
+    private String getMetadataContainerElement()
+    {
+        if (getBlockState().isInLine()) {
+            return "span";
+        } else {
+            return "div";
+        }
+    }
+
+    @Override
+    public void beginMetaData(MetaData metadata)
+    {
+        Map<String, String> attributes = new LinkedHashMap<>();
+
+        for (Map.Entry<String, Object> metadataPair : metadata.getMetaData().entrySet()) {
+            attributes.put(XHTMLXWikiGeneratorListener.METADATA_ATTRIBUTE_PREFIX + metadataPair.getKey(),
+                metadataPair.getValue().toString());
+        }
+
+        attributes.put("class", XHTMLXWikiGeneratorListener.METADATA_CONTAINER_CLASS);
+
+        this.getXHTMLWikiPrinter().printXMLStartElement(getMetadataContainerElement(), attributes);
```

Comment:
We might want to find some way to do that only if the target syntax is annotated html. Probably means something like having several html macro renderers (like we have several html renderers) and choose one based on the rendering context target syntax.

Reply:
> We might want to find some way to do that only if the target syntax is annotated html.

I'm not entirely sure I agree: it would mean that anyone who writes an html macro that contains a `<div data-xwiki-custom-metadata="value">` would have a different behaviour when viewing or editing. And I'm not sure it's something he/she would want.

OTOH I guess that we should check that context in the implem of the new wikimacrocontent to not add the metadata when we're not in edition.

Reply:
> I'm not entirely sure I agree: it would mean that anyone who writes an html macro that contains a <div data-xwiki-custom-metadata="value"> would have a different behaviour when viewing or editing. And I'm not sure it's something he/she would want.

I'm not suggesting to remove stuff from the HTML content here, I'm only referring HTML generated from rendering blocks (which is what this code is doing) and have consistent behavior with other HTML renderers. If you directly put `<div data-xwiki-custom-metadata="value">` in your html macro you will always end up with `<div data-xwiki-custom-metadata="value">` whatever is the context.

>  when we're not in edition

Edition is just one use case so that's really not the right criteria which is why I talked about target syntax.

Reply:
Indeed you're right. I'll check how to do that.